docs/USAGE-GUIDE.md

```diff
@@ -1929,6 +1929,95 @@ s2n-tls will consider all allocated memory available for use, and will attempt t
 
+
+## Using Early Data / 0RTT
+
+TLS1.3 introduced the ability for clients to send data before completing the handshake when using external pre-shared keys or session resumption.
+
+**WARNING:** Early data does not have the same security properties as regular data sent after a successful handshake.
+* It is not forward secret. If the PSK or session resumption secret is compromised, then the early data is also compromised.
+* It is susceptible to replay attacks unless proper precautions are taken. Early data can be captured and successfully resent by an attacker. See https://tools.ietf.org/rfc/rfc8446#appendix-E.5 and https://tools.ietf.org/rfc/rfc8446#section-8 for more details, and ["Adding anti-replay protection"](#adding-anti-replay-protection) for how to implement counter measures.
+
+_**Do not enable early data for your application unless you have understood and mitigated the risks.**_
+
+### Configuring session resumption for early data
+
+To use early data with session tickets, early data must be enabled on a server by setting the maximum early data allowed to a non-zero value with **s2n_config_set_server_max_early_data_size** or **s2n_connection_set_server_max_early_data_size**. The server then begins issuing tickets that support early data, and clients can use early data when they use those tickets.
+
+### Configuring external pre-shared keys for early data
+
+To use early data with pre-shared keys, individual pre-shared keys must support early data. In addition to configuring the maximum early data allowed, each pre-shared key needs an associated cipher suite and if applicable, application protocol. The server only accepts early data if the pre-shared key's associated cipher suite and application protocol match the cipher suite and the application protocol negotiated during the handshake.
+
+The maximum early data allowed and cipher suite can be set with **s2n_psk_configure_early_data**. If the connection will negotiate an application protocol then the expected application protocol can be set with **s2n_psk_set_application_protocol**. 
+
+### Sending early data
+
+To send early data, your application should call **s2n_send_early_data** before it calls **s2n_negotiate**. Like other IO functions, **s2n_send_early_data** can potentially fail repeatedly with a blocking error before it eventually succeeds: see [I/O Functions](#io-functions) for more information.
+
+**s2n_connection_get_remaining_early_data_size** can be called to check how much more early data the client is allowed to send. If **s2n_send_early_data** exceeds the allowed maximum, s2n-tls returns a usage error.
+
+An application can stop calling **s2n_send_early_data** at any time, even if the function has not returned success yet. If **s2n_send_early_data** does return success, the connection is ready to complete the the handshake and begin sending normal data. However, **s2n_send_early_data** can continue to be called to send more early data if desired.
```

Comment:
`If **s2n_send_early_data** does return success, the connection is ready to complete the the handshake`

what happens if s2n_send_early_data returns an error? should one try and call `s2n_negotiate`? in the example below you call `s2n_negotiate` regardless of the status from the first loop.

Reply:
I moved the discussion of error handling to the same paragraph as the discussion of success. 

The example should exit(1) if the first loop produces an error other than S2N_ERR_T_BLOCKED, and that's what it looks like it's doing to me. Is there a mistake in the code?